vfs_handlers/s3_handlers.cc

```diff
@@ -122,292 +122,416 @@ struct AtInit {
-                    // and throw an appropriate exception
-                    string chunkEtag = partResult.getHeader("etag") ;
-                    if(chunkEtag != info.etag)
-                        throw ML::Exception("chunk etag %s not equal to file etag %s: file <%s> has changed during download!!", chunkEtag.c_str(), info.etag.c_str(), object.c_str());
-                    ExcAssert(partResult.body().size() == chunkSize);
-
-                    while (true) {
-                        if (shutdown || lastExc) {
-                            return;
-                        }
-                        if (threadQueue.tryPush(partResult.body())) {
-                            break;
-                        }
-                        else {
-                            std::this_thread::sleep_for(std::chrono::milliseconds(100));
-                        }
-                    }
-                }
+    void handleResponse(unsigned int chunkNr, size_t chunkSize,
+                        S3Api::Response && response,
+                        std::exception_ptr excPtr)
+    {
+        try {
+            if (excPtr) {
+                rethrow_exception(excPtr);
+            }
+
+            if (response.code_ != 200 && response.code_ != 206) {
+                throw ML::Exception("http error "
+                                    + to_string(response.code_)
+                                    + " while getting chunk "
+                                    + response.bodyXmlStr());
             }
-            catch (...) {
-                lastExc = current_exception();
+
+            /* It can sometimes happen that a file changes during download i.e
+               it is being overwritten. Make sure we check for this condition
+               and throw an appropriate exception. */
+            string chunkEtag = response.getHeader("etag");
+            if (chunkEtag != fileInfo.etag) {
+                throw ML::Exception("chunk etag '%s' differs from original"
+                                    " etag '%s' of file '%s'",
+                                    chunkEtag.c_str(), fileInfo.etag.c_str(),
+                                    resource.c_str());
             }
+            ExcAssertEqual(response.body().size(), chunkSize);
+            Chunk & chunk = chunks[chunkNr];
+            chunk.assign(std::move(response.body_));
+        }
+        catch (const std::exception & exc) {
+            excPtrHandler.takeCurrentException();
         }
+        activeRqs--;
+        ML::futex_wake(activeRqs);
+    }
 
-        size_t getChunkSize(unsigned int chunkNbr)
-            const
-        {
-            size_t chunkSize = std::min(baseChunkSize * (1 << (chunkNbr / 2)),
-                                        maxChunkSize);
-            return chunkSize;
+    void handleEtagChange()
+    {
+        auto newInfo = api->getObjectInfo(bucket, resource.substr(1));
```

Comment:
Originally, it was a hack designed to ensure that the reason why the etag could differ was the expected reason: that the file had indeed changed during the download. I think this hack can be removed but it does not cause much harm.